Adding Assertions#assertThatObject to force ObjectAssertions

AHeise opened the pull request:
#### Check List:
* Fixes https://github.com/joel-costigliola/assertj-core/issues/1388
* Unit tests : YES
* Javadoc with a code example (API only) : YES
### 

Currently, proposal for the issue. If it's acceptable, I'd bring the new method also to SoftAssertions etc.

joel-costigliola commented on src/main/java/org/assertj/core/api/Assertions.java:
```diff
@@ -1104,6 +1104,31 @@ public static AtomicLongArrayAssert assertThat(AtomicLongArray actual) {
+   * @param actual the actual value.
+   * @return the created assertion object.
+   */
+  public static <T> ObjectAssert<T> assertThatObject(T actual) {
```
this also needs to be added to `WithAssertion` and `BDDAssertions` but for the latter as `thenObject`

joel-costigliola commented:
@AHeise thanks for the PR, I have just reviewed it and there is still a bit of work to do before it is ready to be merged.

joel-costigliola commented:
Gentle bump, @AHeise I'm wrapping up the work for the next release (3.12.0) in the next weeks, it would be good for my comments to be addressed if you want to have this issue addressed in 3.12.0. It's alright if you are too busy, I'll just move this for 3.13.0.

GaspardPO commented on src/test/java/org/assertj/core/api/Assertions_assertThatObject_Test.java:
```diff
@@ -0,0 +1,61 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2018 the original author or authors.
```
if we really want to have a Copyright in the doc, then it's 2012-2019.

joel-costigliola replied:
we will get there with the latest parent pom: https://github.com/joel-costigliola/assertj-maven-parent-pom/blob/master/pom.xml#L387

AHeise commented:
Thanks for the comments, I will work on that today.

Concerning the year - will that be replaced automatically or should I adjust it in the headers?

AHeise commented:
The Assertions_sync leaves currently quite a bit of methods untested (assertThatCode vs. thenCode). I'm currently thinking of reworking the sync test a bit to work prefix-based (check all methods starting with assertThat and then respectively and assume same postfix).

AHeise commented:
It turned out that I also needed to add the respective methods to the soft assertions. I squashed all 5 additions into one commit.

Like I outlined above, I also reworked the sync test, which is the second commit. However, there were other methods missing, mostly focused on exceptions.
Missing soft and bdd assertions:
`
  <[public static <T> org.assertj.core.api.ThrowableTypeAssert<T> org.assertj.core.api.Assertions.assertThatExceptionOfType(java.lang.Class<? extends T>),
    public static org.assertj.core.api.ThrowableTypeAssert<java.lang.NullPointerException> org.assertj.core.api.Assertions.assertThatNullPointerException(),
    public static org.assertj.core.api.ThrowableTypeAssert<java.lang.IllegalArgumentException> org.assertj.core.api.Assertions.assertThatIllegalArgumentException(),
    public static org.assertj.core.api.ThrowableTypeAssert<java.io.IOException> org.assertj.core.api.Assertions.assertThatIOException(),
    public static org.assertj.core.api.ThrowableTypeAssert<java.lang.IllegalStateException> org.assertj.core.api.Assertions.assertThatIllegalStateException()]>
`

Is this by design? If not we should probably add them as well.

AHeise commented:
@joel-costigliola could you please decide whether the revised sync tests is working for you? We might also create a different ticket for that. That could also be the place where I'd add the assertThatXException to bdd and soft assertions.

joel-costigliola commented:
@AHeise the build is failing with 3 errors https://travis-ci.org/joel-costigliola/assertj-core/jobs/488712156.

Besides fixing this errors, there are a few things to add:

- add `assumeThatObject` in `Assumptions` + a test in `Assumptions_assumeThat_with_various_types_Test`
- add a soft assertions test with `assertThatObject` in `SoftAssertionsTest`
- a similar test BDD soft assertions with `thenObject` in `BDDSoftAssertionsTest`

I'm going to release 3.12.0 be the end to the week, if you can make it by Wednesday then I'll propbably have the time to include feature otherwise no stress it will be in 3.13.0.

AHeise commented:
Hi @joel-costigliola , could you please look at my questions from a week ago? The tests fail because there is inconsistency on the assertThatXException methods. I can fix them but want to confirm that this is not by design and an actual oversight.

joel-costigliola commented:
I confirm it was an oversight, it would be great to fix them but that would require to add the corresponding tests which is quite a bit of work. 

If you have the time, great, otherwise it might be better to create another issue to track that work (which is unrelated to this PR anyway) and adapt your tests to ignore these discrepancies.

Let me know what works best for you.

AHeise commented:
I have added the missing the tests, but I'd like to extract the whole sync refactoring + adding missing methods to a separate issue. I opened #1434 for that. I will rebase the respective commit to a 1434 PR.

My plan would be to finish this ticket for the upcoming release and work on 1434 asap but without any time commitment.

joel-costigliola commented:
One build [failed](https://travis-ci.org/joel-costigliola/assertj-core/jobs/492386083) due to javadoc errors, I believe this is due because `<>` are not HTML escaped.

joel-costigliola commented on src/main/java/org/assertj/core/api/Assertions.java:
```diff
@@ -1104,6 +1104,31 @@ public static AtomicLongArrayAssert assertThat(AtomicLongArray actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<T>` to `&lt;T&gt;`

joel-costigliola commented on src/main/java/org/assertj/core/api/Assumptions.java:
```diff
@@ -835,6 +835,32 @@ public static AtomicLongArrayAssert assumeThat(AtomicLongArray actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<T>` to `&lt;T&gt;`

joel-costigliola commented on src/main/java/org/assertj/core/api/BDDAssertions.java:
```diff
@@ -1068,6 +1068,31 @@ public static AtomicLongArrayAssert then(AtomicLongArray actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented on src/main/java/org/assertj/core/api/Java6AbstractBDDSoftAssertions.java:
```diff
@@ -768,6 +768,31 @@ public ThrowableAssert then(Throwable actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented on src/main/java/org/assertj/core/api/Java6AbstractStandardSoftAssertions.java:
```diff
@@ -756,6 +756,31 @@ public ThrowableAssert assertThat(Throwable actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented on src/main/java/org/assertj/core/api/Java6Assertions.java:
```diff
@@ -1065,6 +1065,31 @@ public static AtomicLongArrayAssert assertThat(AtomicLongArray actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented on src/main/java/org/assertj/core/api/Java6BDDAssertions.java:
```diff
@@ -926,6 +926,31 @@ public static AtomicLongArrayAssert then(AtomicLongArray actual) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented on src/main/java/org/assertj/core/api/WithAssertions.java:
```diff
@@ -2314,6 +2314,31 @@ default OptionalLongAssert assertThat(final OptionalLong optional) {
+   * </p>
+   *
+   * Example :
+   * <pre><code class='java'>class MyList<T> extends AbstractList<T> {
```
escape `<` and `>` to `&lt;` and `&gt;` everywhere in the javadoc

joel-costigliola commented:
Almost there, thanks @AHeise for the work!

AHeise commented:
Okay managed to solve the javadoc issues by using the @code block; hope that's okay. Now all tests are green.

joel-costigliola commented:
Can you check that the javadoc comes up with syntax highlighting for code snippet ?
We use <pre><code> to get syntax highlighting, I'm not sure that works with `@code`

AHeise commented:
It looks differently, I will fix that now.

AHeise commented:
@code actually works well. I just forgot to wrap it in an <pre> tag to get the nice block quote layout. Fixed it and now the code listings in the docs look alike without the need to escape any HTML characters.

For further reading, check 
https://docs.oracle.com/javase/1.5.0/docs/tooldocs/windows/javadoc.html#%7B@code%7D
@code is actually used quite often in assertj javadoc, just not for code listings, where it offers the greatest benefits.

joel-costigliola commented:
Finally integrated! Thanks @AHeise !
